README.md
# ORMP

This repository includes various Python implementations of the Order Recursive Matching Pursuit (ORMP) algorithm for solving the sparse linear regression problem.

## Introduction

Here, we aim to solve various versions of sparse linear regression, also known in the signal processing community as sparse approximation, sparse representation, compressed sensing or signal reconstruction; other algorithms include [matching pursuit](https://en.wikipedia.org/wiki/Matching_pursuit) and its variants, such as orthogonal matching pursuit (OMP), regularized OMP, stagewise OMP, compressed sensinting matching pursuit, and subspace pursuit.

The problem we aim to solve is the following: Given an $n$-dimensional vector $y \in \mathbb{R}^n$ of observed values, and an $n \times p$-dimensional matrix $X$ of regressors, find parameters $\beta \in \mathbb{R}^p$ so that $y = X\beta + \varepsilon$, with $\varepsilon$ as small as possible; this is nothing but linear regression, but we also want to require that $\beta$ has only few non-zero entries.

Specifically, let $S_k = \lbrace \beta \in \mathbb{R}^p \mid \lVert \beta \rVert_0 \leq k \rbrace$ be the set of vectors in $\mathbb{R}^p$ with at most $k$ non-zero entries. We then consider the problem of determining

$$\min_{\beta \in S_k} \lVert X\beta - y \rVert_2,$$

for a fixed allowed number of non-zero entries $k \geq 0$. We also consider the corresponding problem of determining the smallest value of $k$ to achieve a given tolerance $T \geq 0$ on the residual vector $\varepsilon$,

$$\min \lbrace k \in \mathbb{N} \mid \exists \beta \in S_k : \lVert X\beta - y \rVert_2 \leq T \rbrace.$$

We do not solve the problems to optimality but instead provide a greedy heuristic which iteratively adds non-zero entries (hence features to the linear model). This is similar to what happens in matching pursuit, where one iteratively picks the column with maximum inner product with the current residual; here, instead, we choose the column which minimizes the actual objective function (the length of the projection of $y$ on the subspace obtained by adding the column); generally speaking, this can be seen as a more thorough greedy approach than matching pursuit, which will, at least in cases of interest to us, yield better results but generally at the cost of spending more time to get those results.

The solvers come with implementations of the scikit-learn estimator API and therefore can be used as a part of scikit-learn pipelines. We provide an implementation that relies only on NumPy, as well as an implementation that makes use of JAX to be able to make efficient use of any available GPUs/TPUs.

## Installation

(TODO: Not yet true) The package is available on PyPI and can be installed using

```
pip install ormp
```

As the best possible configuration of JAX depends on the system the code will be running on, and may include things like CUDA being readily available when running the solver on a GPU, JAX is not included as a dependency and to make use of the JAX-based solvers, it must be installed separately. Refer to the [JAX installation guide](https://github.com/google/jax/#installation) for possible options. If you only wish to quickly get a feel for the performance of ORMP, you can start by installing it without JAX and using only the already efficient NumPy-based solver.

## Usage

In the example below, we generate a random example with 100 regressors and ask for a fit that uses at most 10 of them.

```python
>>> from ormp import OrderRecursiveMatchingPursuit
>>> from sklearn.datasets import make_regression
>>> from sklearn.preprocessing import normalize
>>> X, y = make_regression(noise=4, random_state=0)
>>> X = normalize(X, norm="l2", axis=0)
>>> reg = OrderRecursiveMatchingPursuit(n_nonzero_coefs=10, fit_intercept=False).fit(X, y)
>>> reg.score(X, y)
0.9991885378269406
>>> reg.predict(X[:1,])
array([-78.68765328])
>>> reg.coef_
array([  0.        ,   0.        ,  92.43353271,   0.        ,
         0.        ,   0.        ,   0.        ,  77.50286865,
         0.        ,   0.        ,   0.        ,   0.        ,
         0.        ,   0.        ,   0.        ,   0.        ,
         0.        ,   0.        ,   0.        ,   0.        ,
         0.        ,   0.        ,   0.        ,   0.        ,
         0.        , 713.07751465,   0.        ,   0.        ,
         0.        ,   0.        ,   0.        ,   0.        ,
         0.        ,   0.        ,   0.        ,   0.        ,
         0.        ,   0.        ,   0.        ,   0.        ,
         0.        , 382.94140625,   0.        ,   0.        ,
       527.96832275,   0.        , 444.06378174,   0.        ,
         0.        ,   0.        ,   0.        ,   0.        ,
         0.        ,   0.        ,   0.        ,   0.        ,
       286.8444519 ,   0.        ,   0.        ,   0.        ,
         0.        ,   0.        , 646.28283691,   0.        ,
         0.        ,   0.        ,   0.        ,   0.        ,
         0.        ,   0.        ,   0.        ,   0.        ,
         0.        ,   0.        ,   0.        ,   0.        ,
         0.        ,   0.        ,   0.        ,   0.        ,
         0.        ,   0.        ,   0.        ,   0.        ,
       648.30822754,   0.        ,   0.        ,   0.        ,
         0.        ,   0.        ,   0.        ,   0.        ,
         0.        ,   0.        ,   0.        ,   0.        ,
         0.        , 141.12867737,   0.        ,   0.        ])
```

### Choosing an implementation

As stated above, we provide both non-JAX and several JAX based solvers. Which one to choose depends on characteristics of the input data, whether one can live with a long first initial compilation time, and whether one will be running the solver on inputs with many different shapes, each of which will require recompilation.

To accommodate these different scenarios, we include separate JAX solvers that are optimized for runtime and compilation time.

For example, when run on a laptop with a GPU, fitting a `(1000, 1000)`-shaped sample while requiring no more than 100 non-zero coefficients, a typical picture is that

* `ormp.impl_np.ormp_numpy` will take 30 ms to run
* `ormp.impl_jax.ormp_fast_runtime_jit` will take 23 seconds to compile, then 3 ms to run,
* `ormp.impl_jax.ormp_fast_compilation_jit` will take 180 ms to compile, then 15 ms to run.

For comparison, `sklearn.linear_model.OrthogonalMatchingPursuit.fit` would take about 80 ms on such an input on such a device.


## Comparison with Orthogonal Matching Pursuit

It is illustrative to compare the behavior of the algorithm with that of Orthogonal Matching Pursuit (OMP), [as available in scikit-learn](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.OrthogonalMatchingPursuit.html#sklearn.linear_model.OrthogonalMatchingPursuit).

As mentioned above, ORMP can be considered as a variation on the iterative approach defining OMP, in that both methods are iterative, but here we spend more time in each iteration, picking the input vector that minimizes the residual rather than simply the input vector with the smallest angle with the residual.

Both algorithms being heuristics, it is not clear a priori how much of a benefit this will give on a given input, if any; it is easy to imagine that for certain classes of inputs, one algorithm is a better choice than the other, and certainly our own work here has been inspired by ORMP greatly outperforming OMP on inputs of interest.

One way to get a bit of an intuition about how the algorithms compare is by running them on random small instances. Here, specifically, we run them on 10,000 inputs consisting of columns sampled uniformly on the unit spheres of various dimensions, and vary the number of desired non-zero coefficients; we then compare the fits and keep track of which algorithm led to the better fit:

|   n_samples |   n_features |   n_nonzero_coefs |   OMP wins |   ORMP wins |   Draw |
|-------------|--------------|-------------------|------------|-------------|--------|
|           3 |            3 |                 2 |          0 |        1342 |   8658 |
|           4 |            3 |                 2 |          0 |         761 |   9239 |
|           4 |            4 |                 2 |          0 |        1248 |   8752 |
|           4 |            4 |                 3 |        262 |        1999 |   7739 |
|           5 |            3 |                 2 |          0 |         513 |   9487 |
|           5 |            4 |                 2 |          0 |         923 |   9077 |
|           5 |            4 |                 3 |        145 |        1355 |   8500 |
|           5 |            5 |                 2 |          0 |        1189 |   8811 |
|           5 |            5 |                 3 |        276 |        2189 |   7535 |
|           5 |            5 |                 4 |        572 |        2674 |   6754 |

Note, for instance, that when `n_nonzero_coefs` is 2, OMP never wins. This is the case since the two methods will always pick the same first vector, and ORMP will then always pick the best of the remaining ones. It is however, not the case that ORMP is always the strictly better choice, and for larger instances, OMP starts winning a significant amount of time.

The code to generate a row in this table is given below:

```python
def random_unit_vector(n):
    a = np.random.normal(size=n)
    return a / np.linalg.norm(a)

def random_instance(n, m):
    X = np.empty((n, m), dtype=np.float64)
    for i in range(m):
        X[:, i] = random_unit_vector(n)
    y = random_unit_vector(n)
    return X, y
        
def counts(n, m, n_nonzero_coefs, iters):
    draw = omp_won = ormp_won = 0
    for _ in range(iters):
        X, y = random_instance(m, n)
        coef1 = OrthogonalMatchingPursuit(n_nonzero_coefs=n_nonzero_coefs, fit_intercept=False).fit(X, y).coef_
        coef2 = OrderRecursiveMatchingPursuit(n_nonzero_coefs=n_nonzero_coefs, fit_intercept=False).fit(X, y).coef_
        if np.allclose(coef1, coef2):
            draw += 1
        elif np.linalg.norm(X @ coef1 - y) < np.linalg.norm(X @ coef2 - y):
            omp_won += 1
        elif np.linalg.norm(X @ coef1 - y) > np.linalg.norm(X @ coef2 - y):
            ormp_won += 1
        else:
            assert False
    return draw, omp_won, ormp_won

counts(5, 5, 4, 10_000)
```

CHANGELOG.md
# Changelog

## [Unreleased]

## [0.0.1] - 2023-07-05

Includes two implementations of the ORMP algorithm based on NumPy and JAX.

[Unreleased]: https://github.com/kvantify/ormp/compare/0.0.1...HEAD
[0.0.1]: https://github.com/kvantify/ormp/releases/tag/0.0.1


src/hotpursuit/sklearn.py
"""Implementation of scikit-learn interface to the solvers."""
from numbers import Integral, Real

import numpy as np
from sklearn.base import MultiOutputMixin, RegressorMixin
from sklearn.linear_model._base import LinearModel
from sklearn.linear_model import LinearRegression
from sklearn.utils._param_validation import (
    Interval,
    StrOptions,
)


class HotPursuit(MultiOutputMixin, RegressorMixin, LinearModel):
    """HOT Pursuit model.

    Parameters
    ----------
    n_nonzero_coefs : int, default=None
        Desired number of non-zero entries in the solution. If None (by
        default) this value is set to 10% of n_features.

    tol : float, default=None
        Maximum norm of the residual. If not None, overrides n_nonzero_coefs.

    fit_intercept : bool, default=True
        Whether to calculate the intercept for this model. If set
        to false, no intercept will be used in calculations
        (i.e. data is expected to be centered).

    greediness : int, default=1
        The number of columns to add in one iteration of the algorithm. Lower
        values will give better results but fits take longer.

    Attributes
    ----------
    coef_ : ndarray of shape (n_features,) or (n_targets, n_features)
        Parameter vector (w in the formula).

    intercept_ : float or ndarray of shape (n_targets,)
        Independent term in decision function.

    Notes
    -----
    TODO: Reference for HOT pursuit documentation.

    Examples
    --------
    >>> from hotpursuit.sklearn import HotPursuit
    >>> from sklearn.datasets import make_regression
    >>> from sklearn.preprocessing import normalize
    >>> X, y = make_regression(noise=4, random_state=0)
    >>> X = normalize(X, norm="l2", axis=0)
    >>> reg = HotPursuit(n_nonzero_coefs=10, fit_intercept=False).fit(X, y)
    >>> reg.score(X, y)
    0.9991885378269406
    >>> reg.predict(X[:1,])
    array([-78.68765328])
    """

    _parameter_constraints: dict = {
        "n_nonzero_coefs": [Interval(Integral, 1, None, closed="left"), None],
        "tol": [Interval(Real, 0, None, closed="left"), None],
        "fit_intercept": ["boolean"],
        "implementation": [StrOptions({"numpy", "jax"})],
        "greediness": [Interval(Integral, 1, None, closed="left")],
    }

    def __init__(
        self, *, n_nonzero_coefs=None, tol=None, fit_intercept=True, greediness=1
    ):
        self.n_nonzero_coefs = n_nonzero_coefs
        self.greediness = greediness
        if tol is not None:
            raise NotImplementedError("tol is not yet implemented")
        if fit_intercept:
            raise NotImplementedError(
                "the only implemented value of fit_intercept is False"
            )

    @staticmethod
    def _add_column(X, A_inv, u, dtype):
        X_prime = np.column_stack((X, u))
        w = X.T @ u
        A_inv_w = A_inv @ w
        # We assume X is normalized so u.T @ u = 1
        one_by_denominator = 1 / (1 - w.T @ A_inv_w)
        A_prime_inv_bottom_right = one_by_denominator
        n = A_inv.shape[0]
        if n == 0:
            A_prime_inv = np.array([[A_prime_inv_bottom_right]])
        else:
            numerator = np.outer(A_inv_w, A_inv_w.T)
            A_prime_inv_top_left = A_inv + numerator * one_by_denominator
            off_diag = -A_inv_w * one_by_denominator
            # The new A⁻¹ is a block matrix whose top left n x n block
            # is the original A⁻¹ matrix; it turns out that creating this
            # block matrix from an empty matrix is faster than using e.g. np.block
            A_prime_inv = np.empty((n + 1, n + 1), dtype=dtype)
            A_prime_inv[:n, :n] = A_prime_inv_top_left
            A_prime_inv[-1, :n] = off_diag
            A_prime_inv[:n, -1] = off_diag.T
            A_prime_inv[-1, -1] = A_prime_inv_bottom_right
        return X_prime, A_prime_inv

    def fit(self, X, y):
        """Fit the model using X, y as training data.

        Parameters
        ----------
        X : array-like of shape (n_samples, n_features)
            Training data.

        y : array-like of shape (n_samples,)
            Target values.

        Returns
        -------
        self : object
            Returns an instance of self.
        """

        X, y = self._validate_data(X, y, multi_output=True, y_numeric=True)
        # On our test cases, explicitly sticking to np.float64 instead of np.float32
        # yields a ~26% overall reduction in time to fit.
        dtype = np.float32

        # Ensure X is normalized
        n = X.shape[0]
        if np.any(np.abs(np.sum(X * X, axis=0) - 1) > 1e-7):
            raise ValueError("X must have normalized columns")

        # Keep track of which column indices have been included so far
        remaining_indices = np.ones(n, dtype=np.bool_)
        X = X.astype(dtype)
        y = y.astype(dtype)
        X_curr = np.empty((n, 0), dtype=dtype)
        A_inv_curr = np.empty((0, 0), dtype=dtype)
        k = self.n_nonzero_coefs
        d = self.greediness
        while remaining_indices.sum() > n - k:
            pre_calc_proj = X_curr.T @ y
            us = X[:, remaining_indices]
            ws = X_curr.T @ us  # TODO: This can be simplified as we know X_prev.T @ us
            A_inv_ws = A_inv_curr @ ws
            w_t_A_inv_ws = np.sum(ws * A_inv_ws, axis=0)
            one_by_denominators = 1 / (1 - w_t_A_inv_ws)

            top_left_static = A_inv_curr
            top_left_dynamic = (
                np.einsum("jk, ik, j->ik", A_inv_ws, A_inv_ws, pre_calc_proj)
                * one_by_denominators
            )
            off_diags = -A_inv_ws * one_by_denominators[np.newaxis, :]

            proj_bottom = us.T @ y
            weight_top_left_static = top_left_static @ pre_calc_proj
            weight_top_left_dynamic = top_left_dynamic
            weight_top_left = (
                weight_top_left_static[:, np.newaxis] + weight_top_left_dynamic
            )
            weight_top_right = off_diags * proj_bottom
            weight_top = weight_top_left + weight_top_right

            weight_bottom_left = np.einsum("ij,i->j", off_diags, pre_calc_proj)
            weight_bottom_right = one_by_denominators * proj_bottom
            weight_bottom = weight_bottom_left + weight_bottom_right

            # The below alternative is almost faster, and gets rid of top_left_dynamic
            # above, which would be nice, but we keep the above for performance:
            #
            # weight_top_small = weight_top_left_static[:,np.newaxis] +\
            #      off_diags * proj_bottom
            # p1 = np.einsum(
            #     'mi, jk, ik, j->mk', X_curr, A_inv_ws,
            #     A_inv_ws, pre_calc_proj, optimize='optimal'
            # ) * one_by_denominators
            # p2 = X_curr @ weight_top_small
            # y_hats = p1 + p2 + us * weight_bottom

            y_hats = X_curr @ weight_top + us * weight_bottom

            diffs = y_hats - y[:, np.newaxis]
            all_evals = np.sum(diffs * diffs, axis=0)
            to_choose = np.argpartition(all_evals, d)[:d]

            evals = {
                p: all_evals[i] for i, p in enumerate(np.where(remaining_indices)[0])
            }
            sorted_evals = sorted(evals.items(), key=lambda x: x[1])
            to_choose = [x for x, _ in sorted_evals[:d]]
            for new_col in to_choose:
                X_curr, A_inv_curr = self._add_column(
                    X_curr, A_inv_curr, X[:, new_col], dtype=dtype
                )
            remaining_indices[to_choose] = 0

        # Build solution
        reg = LinearRegression(fit_intercept=False).fit(X[:, ~remaining_indices], y)
        self.coef_ = np.zeros(X.shape[1])
        self.coef_[~remaining_indices] = reg.coef_
        self.intercept_ = 0.0
        return self
